core/kernel/src/feature_manifest.rs

```diff
@@ -0,0 +1,270 @@
+use serde::Deserialize;
+use std::collections::HashMap;
+
+#[derive(Deserialize, PartialEq, Debug)]
+#[serde(rename_all = "camelCase")]
+#[allow(dead_code)] // Temporary until exposed through C API
+pub struct FeatureManifest {
+    name: Option<String>,
+    author: Option<String>,
+    version: Option<String>,
+    description: Option<String>,
+    thumbnail: Option<String>,
+    thumbnail_color: Option<String>,
+    primary_color: Option<String>,
+    links: Option<HashMap<String, String>>,
+}
+
+#[derive(PartialEq, Debug)]
+pub struct FeatureManifestParsingError {
+    description: String,
+}
+
+// Alias for str, probably will be moved to a centralized place to be reused
+pub type JSONStr = str;
+
+#[allow(dead_code)] // Temporary until exposed through C API
+impl FeatureManifest {
+    pub fn parse(
+        json: &JSONStr,
+        locale: &str,
+    ) -> Result<FeatureManifest, FeatureManifestParsingError> {
+        FullFeatureManifest::try_from(json)
+            .map(|manifest| FeatureManifest::build_feature_manifest(manifest, locale))
+    }
+
+    fn build_feature_manifest(full_manifest: FullFeatureManifest, locale: &str) -> FeatureManifest {
+        let translation = full_manifest
+            .translations
+            .as_ref()
+            .and_then(|unwrapped| unwrapped.get(locale));
+
+        match translation {
+            Some(translation) => {
+                let mut links = full_manifest.links.unwrap_or_default();
+                if let Some(translated_links) = translation.links.clone() {
+                    links.extend(translated_links.into_iter());
+                }
+                FeatureManifest {
+                    name: translation
```

Comment:
Any way to un-DRY this?

Reply:
You mean code like `translation.name.clone().or(full_manifest.name)`? The only thing I am thinking of is having a function to contain the `....clone.or(...)` "logic", but I am not sure it would be better in terms of readability.

Reply:
No, just to avoid three lines of code for every attribute.